File: whatsapp-service-go/internal/agent/protocol.go
package agent

import (
	"crypto/sha256"
	"encoding/hex"
	"time"
)

// Helper para calcular hash SHA-256 (primeiros 10 caracteres) para rastreabilidade de ponta a ponta
func HashQR(qr string) string {
	h := sha256.Sum256([]byte(qr))
	return hex.EncodeToString(h[:])[:10]
}

// Tipos base
type WhatsAppStatus string
type SessionStatus string
type JobStatus string

const (
	StatusStarting       WhatsAppStatus = "STARTING"
	StatusNoSession      WhatsAppStatus = "NO_SESSION"
	StatusWaitingQR      WhatsAppStatus = "WAITING_QR"
	StatusAuthenticating WhatsAppStatus = "AUTHENTICATING"
	StatusConnected      WhatsAppStatus = "CONNECTED"
	StatusDisconnected   WhatsAppStatus = "DISCONNECTED"
	StatusReconnecting   WhatsAppStatus = "RECONNECTING"
	StatusAuthRequired   WhatsAppStatus = "AUTH_REQUIRED"
	StatusProtocolError  WhatsAppStatus = "PROTOCOL_ERROR"
	StatusError          WhatsAppStatus = "ERROR"

	SessionNone  SessionStatus = "NONE"
	SessionValid SessionStatus = "VALID"

	JobClaimed    JobStatus = "CLAIMED"
	JobProcessing JobStatus = "PROCESSING"
	JobSent       JobStatus = "SENT"
	JobFailed     JobStatus = "FAILED"
)

type BaseMessage struct {
	Type      string `json:"type"`
	Version   int    `json:"version"`
	RequestID string `json:"requestId"`
	Timestamp string `json:"timestamp"`
}

// ─── Agent → DO ──────────────────────────────────────────────────────────────

type RegisterAgentMsg struct {
	BaseMessage
	AgentID      string   `json:"agentId"`
	AgentVersion string   `json:"agentVersion"`
	Capabilities []string `json:"capabilities"`
}

type HeartbeatMsg struct {
	BaseMessage
	AgentID        string         `json:"agentId"`
	WhatsAppStatus WhatsAppStatus `json:"whatsappStatus"`
	SessionStatus  SessionStatus  `json:"sessionStatus"`
}

type StatusUpdateMsg struct {
	BaseMessage
	AgentID        string         `json:"agentId"`
	WhatsAppStatus WhatsAppStatus `json:"whatsappStatus"`
	SessionStatus  SessionStatus  `json:"sessionStatus"`
	LastError      string         `json:"lastError,omitempty"`
	AgentVersion   string         `json:"agentVersion,omitempty"`
}

type QrUpdateMsg struct {
	BaseMessage
	AgentID     string `json:"agentId"`
	QRCode      string `json:"qrCode"`
	QRVersion   int    `json:"qrVersion,omitempty"`
	GeneratedAt string `json:"generatedAt,omitempty"`
	ExpiresAt   string `json:"expiresAt"`
}

type SessionUpdateMsg struct {
	BaseMessage
	AgentID       string        `json:"agentId"`
	SessionStatus SessionStatus `json:"sessionStatus"`
}

type JobAckMsg struct {
	BaseMessage
	AgentID string    `json:"agentId"`
	JobID   string    `json:"jobId"`
	Status  JobStatus `json:"status"`
	SentAt  string    `json:"sentAt,omitempty"`
}

type JobFailedMsg struct {
	BaseMessage
	AgentID   string `json:"agentId"`
	JobID     string `json:"jobId"`
	Error     string `json:"error"`
	Attempt   int    `json:"attempt"`
	Retryable bool   `json:"retryable"`
}

type ChannelDestination struct {
	JID  string `json:"jid"`
	Name string `json:"name"`
	Type string `json:"type"`
}

type ChannelsUpdateMsg struct {
	BaseMessage
	AgentID  string               `json:"agentId"`
	Channels []ChannelDestination `json:"channels"`
}

// ─── DO → Agent ──────────────────────────────────────────────────────────────

// Generic struct to deserialize incoming DO messages and determine their Type
type IncomingMessage struct {
	Type string `json:"type"`
}

type SendJobMsg struct {
	BaseMessage
	JobID        string `json:"jobId"`
	StudyID      string `json:"studyId"`
	ChannelID    string `json:"channelId"`
	RecipientJID string `json:"recipientJid"`
	Content      string `json:"content"`
	MediaUrl     string `json:"mediaUrl,omitempty"`
}

type CancelJobMsg struct {
	BaseMessage
	JobID string `json:"jobId"`
}

type SyncResponseMsg struct {
	BaseMessage
	PendingJobs []SendJobMsg `json:"pendingJobs"`
}

// Helper to fill BaseMessage fields
func fillBase(msgType string, idGenerator func() string) BaseMessage {
	return BaseMessage{
		Type:      msgType,
		Version:   1,
		RequestID: idGenerator(),
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
}

File: cloudflare-worker/src/protocol.ts
/**
 * Teleios Agent Protocol v1
 * Mensagens trafegam como JSON sobre WebSocket.
 * O Agent sempre inicia a conexão (nunca o servidor).
 */

// ─── Tipos base ──────────────────────────────────────────────────────────────

export type AgentStatus = 'ONLINE' | 'STALE' | 'OFFLINE';

export type WhatsAppStatus =
  | 'STARTING'
  | 'NO_SESSION'
  | 'WAITING_QR'
  | 'AUTHENTICATING'
  | 'CONNECTED'
  | 'DISCONNECTED'
  | 'RECONNECTING'
  | 'AUTH_REQUIRED'
  | 'PROTOCOL_ERROR'
  | 'ERROR';

export type SessionStatus = 'NONE' | 'VALID';

export type JobStatus =
  | 'PENDING'
  | 'READY'
  | 'CLAIMED'
  | 'PROCESSING'
  | 'SENT'
  | 'FAILED'
  | 'RETRYING'
  | 'CANCELLED';

// ─── Envelope base ───────────────────────────────────────────────────────────

export interface BaseMessage {
  type: string;
  version: 1;
  requestId: string;
  timestamp: string; // ISO8601
}

// ─── Agent → DO ──────────────────────────────────────────────────────────────

export interface RegisterAgentMsg extends BaseMessage {
  type: 'REGISTER_AGENT';
  agentId: string;
  agentVersion: string;
  capabilities: string[];
}

export interface HeartbeatMsg extends BaseMessage {
  type: 'HEARTBEAT';
  agentId: string;
  whatsappStatus: WhatsAppStatus;
  sessionStatus: SessionStatus;
}

export interface StatusUpdateMsg extends BaseMessage {
  type: 'STATUS_UPDATE';
  agentId: string;
  whatsappStatus: WhatsAppStatus;
  sessionStatus: SessionStatus;
  lastError?: string;
  agentVersion?: string;
}

export interface QrUpdateMsg extends BaseMessage {
  type: 'QR_UPDATE';
  agentId: string;
  qrCode: string;      // string base do QR (para renderizar com qrcode.react)
  qrVersion?: number;
  generatedAt?: string;
  expiresAt: string;   // ISO8601
}

export interface SessionUpdateMsg extends BaseMessage {
  type: 'SESSION_UPDATE';
  agentId: string;
  sessionStatus: SessionStatus;
}

export interface JobAckMsg extends BaseMessage {
  type: 'JOB_ACK';
  agentId: string;
  jobId: string;
  status: 'CLAIMED' | 'PROCESSING' | 'SENT';
  sentAt?: string;
}

export interface JobFailedMsg extends BaseMessage {
  type: 'JOB_FAILED';
  agentId: string;
  jobId: string;
  error: string;
  attempt: number;
  retryable: boolean;
}

export interface ChannelItem {
  jid: string;
  name: string;
  type: string;
}

export interface ChannelsUpdateMsg extends BaseMessage {
  type: 'CHANNELS_UPDATE';
  agentId: string;
  channels: ChannelItem[];
}

export type AgentToDoMessage =
  | RegisterAgentMsg
  | HeartbeatMsg
  | StatusUpdateMsg
  | QrUpdateMsg
  | SessionUpdateMsg
  | JobAckMsg
  | JobFailedMsg
  | ChannelsUpdateMsg;

// ─── DO → Agent ──────────────────────────────────────────────────────────────

export interface AgentAcceptedMsg extends BaseMessage {
  type: 'AGENT_ACCEPTED';
  serverId: string;
  serverTime: string;
}

export interface PingMsg extends BaseMessage {
  type: 'PING';
}

export interface SendJobMsg extends BaseMessage {
  type: 'SEND_JOB';
  jobId: string;
  studyId: string;
  channelId: string;
  recipientJid: string;  // ex: "5511999998888@s.whatsapp.net"
  content: string;
  mediaUrl?: string;
}

export interface CancelJobMsg extends BaseMessage {
  type: 'CANCEL_JOB';
  jobId: string;
}

export interface PauseMsg extends BaseMessage {
  type: 'PAUSE';
}

export interface ResumeMsg extends BaseMessage {
  type: 'RESUME';
}

export interface SyncResponseMsg extends BaseMessage {
  type: 'SYNC_RESPONSE';
  pendingJobs: SendJobMsg[];
}

export interface RestartWaMsg extends BaseMessage {
  type: 'RESTART_WA';
}

/** Enviado ao agent para iniciar primeira conexão/pareamento sem descartar sessão. */
export interface ConnectWaMsg extends BaseMessage {
  type: 'CONNECT_WA';
}

export interface SyncGroupsMsg extends BaseMessage {
  type: 'SYNC_GROUPS';
}

export type DoToAgentMessage =
  | AgentAcceptedMsg
  | PingMsg
  | SendJobMsg
  | CancelJobMsg
  | PauseMsg
  | ResumeMsg
  | RestartWaMsg
  | ConnectWaMsg
  | SyncGroupsMsg
  | SyncResponseMsg;


// ─── DO → Admin clients (broadcast) ─────────────────────────────────────────

export interface AdminStatusMsg extends BaseMessage {
  type: 'ADMIN_STATUS';
  agentStatus: AgentStatus;
  whatsappStatus: WhatsAppStatus;
  sessionStatus: SessionStatus;
  lastSeen: number;   // unix ms
  lastError?: string;
  qrCode?: string | null;
  qrExpiresAt?: string | null;
  qrVersion?: number;
}

export interface AdminQrMsg extends BaseMessage {
  type: 'ADMIN_QR';
  qrCode: string;
  expiresAt: string;
  qrVersion?: number;
  qrHash?: string;
  generatedAt?: string;
}

export interface AdminJobUpdateMsg extends BaseMessage {
  type: 'ADMIN_JOB_UPDATE';
  jobId: string;
  status: JobStatus;
  sentAt?: string;
  error?: string;
}

export type AdminBroadcastMessage = AdminStatusMsg | AdminQrMsg | AdminJobUpdateMsg;

// ─── Tipos de dados persistidos ──────────────────────────────────────────────

export interface Job {
  id: string;
  studyId: string;
  channelId: string;
  agentId: string;
  recipientJid: string;
  content: string;
  mediaUrl?: string;
  scheduledAt: string | null;
  status: JobStatus;
  attempts: number;
  lastError?: string;
  createdAt: string;
  updatedAt: string;
  claimedAt?: string;
  sentAt?: string;
}

export interface Channel {
  id: string;
  name: string;
  whatsappJid: string;
  agentId: string;
  active: boolean;
  createdAt: string;
  updatedAt: string;
}

// ─── Helper ──────────────────────────────────────────────────────────────────

export function makeId(prefix: string): string {
  return `${prefix}_${Date.now()}_${Math.random().toString(36).slice(2, 10)}`;
}

export function makeEnvelope<T extends { type: string }>(
  msg: T
): T & Pick<BaseMessage, 'version' | 'requestId' | 'timestamp'> {
  return {
    version: 1,
    requestId: makeId('req'),
    timestamp: new Date().toISOString(),
    ...msg,
  };
}
